### PR TITLE
WIP:Review

### PR DESCRIPTION
**Pull request by teathedev**

### CODE REVIEW

File: api/authentication.go [hunk 1]
```diff
@@ -84,21 +85,42 @@ func userLogin(c *fiber.Ctx) error {
...
+	token, err := app.CreateJWT(dbUser.ID)
+	if err != nil {
+		return c.Status(400).JSON(err.Error())
+	}
+	return c.Status(200).JSON(token)
+
+}
+func me(c *fiber.Ctx) error {
 
+	tokenString := c.Params("token")
```
Comment (teathedev) [inline api/authentication.go]: get token from `Headers` use something like `X-Authorization` header

File: api/http.go [hunk 2]
```diff
@@ -9,17 +9,14 @@ import (
 var (
 	App     *fiber.App   = fiber.New()
 	UserApi fiber.Router = App.Group("/user")
+	AuthApi fiber.Router = App.Group("/auth")
 )
 
 func ListenPort() {
 
 	err := App.Listen(":3000")
```
Comment (teathedev) [inline api/http.go]: you can do it like this;
```go
if err := App.Listen(":3000"); err != nil {
    logger.Fatal("Error: <?>", err)
}
```

File: app/authentication.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+		logger.Error("Login Error: <?>", err)
+		return err
+	}
+
+	return nil
+}
+
+func CreateJWT(id string) (string, error) {
+	//logine de ek olarak tek kullanılımlık refresh token uret 1-2 saat veya 1 gün
+	//bunlar bir keyle şifreleniyor +jwt tokenle olacak bicimde // tek bir string dönüyorsun bunu da base64lemen dönüyor
```
Comment (teathedev) [inline app/authentication.go]: remove your notes

File: app/authentication.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+import (
+	"emreddit/db"
+	"emreddit/logger"
+	"errors"
+	"time"
+
+	"github.com/golang-jwt/jwt/v5"
+)
+
+var KEY = "THIS IS A KEY"
```
Comment (teathedev) [inline app/authentication.go]: you shouldn't keep things like this key as an hardcoded. You can move these kind of values into `env variables`.

Checkout this package: https://github.com/go-nv/goenv

Comment (teathedev) [reply app/authentication.go]: same thing applies the things like database options, application port (which you gave fiber to listen)

File: app/authentication.go [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+	}
+
+	return nil
+}
+
+func CreateJWT(id string) (string, error) {
+	//logine de ek olarak tek kullanılımlık refresh token uret 1-2 saat veya 1 gün
+	//bunlar bir keyle şifreleniyor +jwt tokenle olacak bicimde // tek bir string dönüyorsun bunu da base64lemen dönüyor
+	token := jwt.NewWithClaims(jwt.SigningMethodHS256, jwt.MapClaims{
+		"exp":     time.Now().Add(time.Hour).Unix(), //15 dk üret
```
Comment (teathedev) [inline app/authentication.go]: update

File: app/authentication.go [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+		"iat":     time.Now().Unix(),
+		"subject": id})
```
Comment (teathedev) [inline app/authentication.go]: new line before `})`

File: app/authentication.go [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+	if err != nil {
+		logger.Error("JWT Generate Error:<?>", err)
+		return "", err
+	}
+	logger.Info(token.Header)
+	//base64le encode
+	return ss, nil
+} //jwt token ise 401 // 401
+func EncryptedToken(id string) {
```
Comment (teathedev) [inline app/authentication.go]: `EncryptToken`

File: migration.go [hunk 8]
```diff
@@ -1,7 +1,8 @@
 package main
 
-import "emreddit/migrate"
+import "emreddit/migration_manager"
 
 func main() {
-	migrate.RunUp()
+	//migration_manager.RunUp()
+	migration_manager.RunDown()
```
Comment (teathedev) [inline migration.go]: ?

File: migration_manager/create_mig.go [hunk 9]
```diff
@@ -9,33 +9,31 @@ import (
 )
 
 type Migration struct {
-	Name           string
-	Timestamp      string
-	TableName      string
-	SuccessInitLog string
+	Name      string `gorm:"column:name;not null;"`
+	Timestamp string
+	TableName string `gorm:"column:name;not null;"`
```
Comment (teathedev) [inline migration_manager/create_mig.go]: why there is two different key with same column name?

File: migration_manager/create_mig.go [hunk 10]
```diff
@@ -9,33 +9,31 @@ import (
...
 	if err != nil {
 		//panic(err)
 		logger.Error(err.Error())
 	}
 
-	File, err := os.Create(fileName) //creating file
-
+	File, err := os.Create(fileName)
 	if err != nil {
 		logger.Error(err)
```
Comment (teathedev) [inline migration_manager/create_mig.go]: it should be `Fatal`

File: api/authentication.go (hunk 1 above)
Comment (ercnbayer) [reply api/authentication.go]:
```func me(c *fiber.Ctx) error {

	var tokenString = c.Get("X-Auth-Token", "null")
	id, err := app.ParseJWT(tokenString)
	if err != nil {
		logger.Error("JWT Token Error:<?>", err)
		return c.Status(400).JSON(err.Error())
	}

	user, err := app.GetUser(id)

	if err != nil {
		logger.Error("Get User Error:<?>", err)
		return c.Status(400).JSON(err.Error())
	}
	return c.Status(200).JSON(&user)
}
is it supposed to be like this

Comment (ercnbayer) [reply api/authentication.go]: and why it is needed

File: api/authentication.go [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+		return c.Status(401).JSON(err.Error())
+	}
+
+	userTokens := app.SessionToken{AccessToken: accessToken, RefreshToken: refreshToken}
+	bytes, err := app.JSONToBytes(&userTokens)
+
+	if err != nil {
+		return c.Status(400).JSON(err.Error())
+	}
+	encoded_token, err := app.EncryptToken(bytes)
```
Comment (teathedev) [inline api/authentication.go]: All of these logic should be in the app layer not api layer

File: api/authentication.go [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+	}
+
+	if err := validator.Validate.Struct(&user); err != nil { //validating struct
+
+		return c.Status(400).JSON(err.Error())
+	}
+
+	var dbUser db.UserEntity
+
+	mapUserLoginPayloadToDbUser(&user, &dbUser) //maping user to db obj
```
Comment (teathedev) [inline api/authentication.go:86]: this too

File: api/authentication.go [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+
+	if err != nil {
+		return c.Status(400).JSON(err.Error())
+	}
+
+	return c.Status(200).JSON(encoded_token)
+
+}
+
+func getAccessToken(c *fiber.Ctx) error {
```
Comment (teathedev) [inline api/authentication.go]: should be `refreshToken`

File: api/authentication.go [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+		return c.Status(401).JSON(err.Error())
+	}
+
+	userTokens := app.SessionToken{AccessToken: accessToken, RefreshToken: refreshToken}
+	bytes, err := app.JSONToBytes(&userTokens)
+
+	if err != nil {
+		return c.Status(400).JSON(err.Error())
+	}
+	encoded_token, err := app.EncryptToken(bytes)
```
Comment (teathedev) [inline api/authentication.go]: should be in app layer

File: app/authentication.go [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+
+func CheckIfTokenValid(refresh_token string) (string, error) {
+
+	token, err := db.ReadToken(refresh_token)
+
+	if err != nil {
+		return "", err
+	}
+
+	if token.ExpireDate.Compare(time.Now()) < 0 || !token.IsUsed {
```
Comment (teathedev) [inline app/authentication.go]: you don't need to do that in here. You can add the database query time filter

File: cmd/migration-cmd/migManagerCmd.go [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+
+	// Here you will define your flags and configuration settings.
+
+	// Cobra supports Persistent Flags which will work for this command
+	// and all subcommands, e.g.:
+	// migManagerCmdCmd.PersistentFlags().String("foo", "", "A help for foo")
+
+	// Cobra supports local flags which will only run when this command
+	// is called directly, e.g.:
+	// migManagerCmdCmd.Flags().BoolP("toggle", "t", false, "Help message for toggle")
```
Comment (teathedev) [inline cmd/migration-cmd/migManagerCmd.go]: clean the comments

File: migration/20240426195151-refresh-token.go [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+
+import (
+	"emreddit/db"
+	"time"
+)
+
+type RefreshToken20240426195151 struct {
+	ID         string    `gorm:"type:uuid;default:gen_random_uuid();primaryKey"`
+	IsUsed     bool      `gorm:"column:Is_Used;not null;default:false"`
+	ExpireDate time.Time `gorm:"column:Expire_Date;"`
```
Comment (teathedev) [inline migration/20240426195151-refresh-token.go]: column names should be lowercase as before

File: api/authentication.go (hunk 11 above)
Comment (ercnbayer) [reply api/authentication.go:111]: i think that should be two functions app.Userlogin and the rest of the creating token process

File: README.md [hunk 18]
```diff
@@ -8,7 +8,10 @@ for running backend
...
  
  for creating migration 
- ```go run create-migration.go -name {migrationName}```
+ ```go run emreddit_cmd.go create {migrationName}```
  
- for running migrations
- ```go run migration.go```
\ No newline at end of file
+ for running migrations up
+ ```go run emreddit_cmd.go up```
```
Comment (teathedev) [inline README.md]: these namings is not needed. you can convert back these. but you can also add a makefile command.

**Pull request merged**